### PR TITLE
[WIP. Don't merge] Resilience to error in packet decoding

### PR DESCRIPTION
**Pull request by 197g**
Addresses #281. Work in progress. Addresses DOS by errors in the decoding step by establishing an error ring-buffer associated to the ethernet interface and socket type. When some decoding step fails, the error is put into the most specific error buffer available but otherwise ignored. This ensures that faulty packets *can not* affect further packet processing. **This is a work in progress, intended for an early preview of ideas. As such, it may be incomplete, not compile at times and/or be buggy. I may possibly rebase or force-push the branch at will.**

A second step of the mechanism design rework will try to rebalance the recv/send inbalance in case of  a saturated `phy`. As layed out in the issue report, a full receive buffer can adversely affect send performance by blocking further sends. The error from the full buffer should break the receive part of the loop but it's not completely clear yet how to handle finally handle it. One possibility is putting it into the ring buffer or the ethernet iface, but this of course creates some friction with the return value a caller expects.

## Alternatives

Callbacks have been evaluated as inferior for major reasons:
* It's missing contextual information resulting from partial decoding. A failed udp decoding could for example still access and log the frames around itself. Re-decoding within the error callback is both inefficient, creates ambiguity and is awkward as it is effectively invoking a processing step that it expects to fail. This seems highly unintuitive.
* Invoking the callback *during* polling means that constraints such as terminiation, timing, resource usage, ... need to be fulfilled by arbitrary user code. Without an extremely extensive documentation, it remains unclear how one could expect the programmer of the callback to know the internals enough to make good engineering decisions about its design.
* The callback must fight the borrow checker if it wants to inspect any state such as the socket on which the error originated or the `phy` interface. Making error inspection a separate call makes this cleaner, less constrained and more efficient.
* Ignoring most errors, as many usages likely want to, would become a no-op for the caller.

## TODO

- [ ] Cleanup some of the changes that were made just to quickly iterate over alternative designs for this feature. This is still ongoing

- [ ] Make the code compile under a subset of features such as (`socket-udp`, `proto-ipv4`). This is actually an issue with the current library and I don't see any guide on which feature set should be supported other than `all` and `default`.

- [ ] Implement error buffers for more than `udp` and decide on a strategy to handle left-over errors that don't belong to any socket.

 - [ ] Evaluate and choose the strategy for recv/send imbalance.

- [ ] Implement tests to simulate and ensure resilience against the DOS potential presented in #281.

### CODE REVIEW

**Comment by whitequark** [comment]
It's an interesting approach. It is a bit awkward, and I don't like that. However, I agree that it seems like our best way going forward, so let's use it.

I think your lifetime changes would need to be a part of a separate PR. They change the public interface as well. I haven't looked at them closely yet as it's hard to follow here.

**Comment by HeroicKatora** [comment]
Yes, it's all a bit raw and several changes are mixed together. I plan to split this up into several independent chunks once I have determined the overall usability.

**Comment by HeroicKatora** [comment]
@whitequark Cleanly rebased only the error buffer related changes. Note that the lifetime within `iface` and `udp` is currently linked to other attributes as I wasn't sure what the original motivation was for having some of them overlap and some of them exposed in the generic parameters. Some feedback on this would be helpful. And, this still needs to be performed for the other socket types to ensure that really no error is generated. I finally may want to remove the `Result<_, Error>` part from the inner methods to make this explicit.

**Comment by m-labs-homu** [comment]
:umbrella: The latest upstream changes (presumably 7c606acdd75853ebc8cf2ece79e7498b7b285a08) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by HeroicKatora** [comment]
Just as a head up: The require changes are far too extensive to justify working on for me. I'm going to abandon this PR for good.

**Comment by whitequark** [comment]
I see. That's a bit unfortunate.
